core/touch_gesture.js

```diff
@@ -0,0 +1,287 @@
+ * events.  "End" refers to touchend, mouseup, and pointerend events.
+ */
+
+/**
+ * Class for one gesture.
+ * @param {!Event} e The event that kicked off this gesture.
+ * @param {!Blockly.WorkspaceSvg} creatorWorkspace The workspace that created
+ *     this gesture and has a reference to it.
+ * @constructor
+ */
+Blockly.TouchGesture = function(e, creatorWorkspace) {
+  Blockly.TouchGesture.superClass_.constructor.call(this, e,
+    creatorWorkspace);
+
+  /**
+   * Boolean for whether or not this gesture is a multi-touch gesture.
+   * @type {boolean}
+   * @private
+   */
+  this.isMultiTouch_ = false;
+
+  /**
+   * A map of cached points used for tracking multi-touch gestures.
+   * @type {Object<number|string, goog.math.Coordinate>}
+   * @private
+   */
+  this.cachedPoints_ = {};
+
+  /**
+   * A scale value tracking the previous gesture scale.
+   * @type {number}
+   * @private
+   */
+  this.previousScale_ = 0;
+
+  /**
+   * The starting distance between two touch points.
+   * @type {number}
+   * @private
+   */
+  this.startDistance_ = 0;
+
+  /**
+   * A handle to use to unbind the second touch start or pointer down listener
+   * at the end of a drag. Opaque data returned from Blockly.bindEventWithChecks_.
+   * @type {Array.<!Array>}
+   * @private
+   */
+  this.onStartWrapper_ = null;
+
+  /**
+   * A cache of workspace metrics containing size and position metrics of the workspace.
+   * @type {Object}
+   * @private
+   */
+  this.metrics_ = null;
+};
+goog.inherits(Blockly.TouchGesture, Blockly.Gesture);
+
+/**
+ * Start a gesture: update the workspace to indicate that a gesture is in
+ * progress and bind mousemove and mouseup handlers.
+ * @param {!Event} e A mouse down, touch start or pointer down event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.doStart = function(e) {
+  Blockly.TouchGesture.superClass_.doStart.call(this, e);
+  if (Blockly.Touch.isTouchEvent(e)) {
+    this.handleTouchStart(e);
+  }
+};
+
+/**
+ * Bind gesture events.
+ * Overriding the gesture definition of this function, binding the same
+ * functions for onMoveWrapper_ and onUpWrapper_ but passing opt_noCaptureIdentifier.
+ * In addition, binding a second mouse down event to detect multi-touch events.
+ * @param {!Event} e A mouse down or touch start event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.bindMouseEvents = function(e) {
+  this.onStartWrapper_ = Blockly.bindEventWithChecks_(
+    document, 'mousedown', null, this.handleStart.bind(this), /*opt_noCaptureIdentifier*/ true);
+  this.onMoveWrapper_ = Blockly.bindEventWithChecks_(
+    document, 'mousemove', null, this.handleMove.bind(this), /*opt_noCaptureIdentifier*/ true);
+  this.onUpWrapper_ = Blockly.bindEventWithChecks_(
+    document, 'mouseup', null, this.handleUp.bind(this), /*opt_noCaptureIdentifier*/ true);
+
+  e.preventDefault();
+};
+
+/**
+ * Handle a mouse down, touch move, or pointer move event.
```

Comment:
Update comment and param: handle a mouse down, touch start or pointer down event.

Reply:
Ah nice catch.